### PR TITLE
mgr/dashboard: Process password complexity checks immediately

### PR DESCRIPTION
**Pull request by votdev**
- Add api/user/validate_password endpoint to check if a password meets the password policy.
- Add Angular async validator to check if the entered password meets the policy.

Fixes: https://tracker.ceph.com/issues/43088

Signed-off-by: Volker Theile <vtheile@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by votdev** [comment]
jenkins test make check

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test make check

**Comment by votdev** [comment]
jenkins test make check arm64

File: src/pybind/mgr/dashboard/controllers/orchestrator.py [hunk 1]
```diff
@@ -2,7 +2,11 @@
 from __future__ import absolute_import
 
 import cherrypy
-from ceph.deployment.drive_group import DriveGroupSpec, DriveGroupValidationError
+
+try:
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/controllers/orchestrator.py]: is this change related to this PR?

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts [hunk 2]
```diff
@@ -59,39 +60,40 @@ export class UserFormComponent implements OnInit {
...
     this.resource = this.i18n('user');
     this.createForm();
     this.messages = new SelectMessages({ empty: this.i18n('There are no roles.') }, this.i18n);
   }
 
   createForm() {
-    this.requiredPasswordRulesMessage = this.userChangePasswordService.getPasswordRulesMessage();
-    this.userForm = new CdFormGroup(
+    this.passwordPolicyHelpText = this.passwordPolicyService.getHelpText();
+    this.userForm = this.formBuilder.group(
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts]: Any reason for this change? This will remove access to methods available in `CdFormGroup`.

File: src/pybind/mgr/dashboard/controllers/user.py [hunk 3]
```diff
@@ -113,6 +123,35 @@ def set(self, username, password=None, name=None, email=None, roles=None,
...
+        :param username: The name of the user (optional).
+        :param old_password: The old password (optional).
+        :return: An object with the properties valid, credits and valuation.
+          'credits' contains the password complexity credits and
+          'valuation' the textual summary of the validation.
+        """
+        result = {'valid': False, 'credits': 0, 'valuation': None}
+        try:
+            result['credits'] = validate_password_policy(password, username, old_password)
+            if result['credits'] < 15:
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/controllers/user.py]: I would suggesting moving this block into a new method in `password-policy.service.ts` (or adapt the current method, since it is already checking for the `credits` value).
Then you could call it in `user-form.component.ts:88` and `user-password-form.component.ts:83`.
This way we could translate the valuation strings.

File: src/pybind/mgr/dashboard/controllers/orchestrator.py (hunk 1 above)
Comment (votdev) [reply src/pybind/mgr/dashboard/controllers/orchestrator.py]: No, but i got lint errors without this change.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts (hunk 2 above)
Comment (votdev) [reply src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts]: > Any reason for this change?

I was not able to get async validators running using the old code.

> This will remove access to methods available in CdFormGroup.

Yes correct, but they are not used here.

File: src/pybind/mgr/dashboard/controllers/user.py (hunk 3 above)
Comment (votdev) [reply src/pybind/mgr/dashboard/controllers/user.py]: Hmmm, IMO this is an strong indication that our i18n system is missing something important here. When we do validations in the backend (what is also planned in the future) and forwarding text to the frontend, then it is absolutely important that we can translate text from the backend, too. Currently failed password policies will raise an exception with an explanation text, this text is forwarded to the frontend and displayed below the input field. Those error messages won't and can't be translated in the moment, too. IMO it would not make sense to relocate the valuation of the password policy check to the frontend only to be able to translate it. Imagine, what happens when the REST API endpoint to validate a password against the policy is called without processing the response in the WebUI; the textual valuation of the password is missed. Because of that the valuation MUST be done in the backend.

Comment (votdev) [reply src/pybind/mgr/dashboard/controllers/user.py]: We will surely run into the same situation sooner or later with other things. The processing and validation of user input becomes more and more complex and we can not do every validation in the frontend. Forwarding textual validation information from the backend to the frontend is somewhat important that we need a solution for that. Refactoring this PR will only schedule the generic problem of our i18n implementation.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/password-policy.service.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+   * Helper function to map password policy credits to a CSS class.
+   * @param credits The password policy credits.
+   * @return The name of the CSS class.
+   */
+  mapCreditsToCssClass(credits: number): string {
+    let result = 'passwordStrengthLevel4';
+    if (credits < 10) {
+      result = 'passwordStrengthLevel0';
+    } else if (credits < 15) {
+      result = 'passwordStrengthLevel1';
```
[the remark this answers is not in the document]
Comment (votdev) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/password-policy.service.ts]: The code is smaller, but ways more complicated to read. That's why I leave it as it is.

File: src/pybind/mgr/dashboard/controllers/user.py [hunk 5]
```diff
@@ -113,6 +123,35 @@ def set(self, username, password=None, name=None, email=None, roles=None,
...
+                result['valuation'] = 'Weak'
+            elif result['credits'] < 20:
+                result['valuation'] = 'OK'
```
[the remark this answers is not in the document]
Comment (votdev) [reply src/pybind/mgr/dashboard/controllers/user.py]: IMO your suggestions will reduce code readability and will increase the complexity, thus i won't apply them.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/password-policy.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (votdev) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/password-policy.service.ts]: I've renamed the CSS classes in the latest commit.

File: src/pybind/mgr/dashboard/controllers/user.py (hunk 3 above)
Comment (LenzGr) [reply src/pybind/mgr/dashboard/controllers/user.py]: Per our discussion during today's standup, we'll accept that these messages are being returned in English only for the time being (as it's already the case in other places, e.g. the Ceph health output or the configuration settings). Going forward (post-Octopus), the Dashboard backend should return localized strings where applicable - this is tracked in [issue#43252](https://tracker.ceph.com/issues/43252).

**Comment by votdev** [comment]
jenkins test dashboard backend

**Comment by votdev** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts (hunk 2 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts]: Didn't notice `formBuilder` was using `CdFormBuilder `.

File: src/pybind/mgr/dashboard/controllers/user.py [hunk 6]
```diff
@@ -113,6 +123,35 @@ def set(self, username, password=None, name=None, email=None, roles=None,
         mgr.ACCESS_CTRL_DB.save()
         return User._user_to_dict(user)
 
+    @Endpoint('POST')
+    @CreatePermission
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/controllers/user.py]: We can't limit the permissions of who calls this, otherwise a read only user won't be able to change password.
```suggestion
```

File: src/pybind/mgr/dashboard/controllers/user.py [hunk 7]
```diff
@@ -113,6 +123,35 @@ def set(self, username, password=None, name=None, email=None, roles=None,
...
+    @UpdatePermission
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/controllers/user.py]:
```suggestion
```

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test make check arm64

**Comment by s0nea** [comment]
jenkins test make check

**Comment by bk201** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test make check arm64

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test dashboard backend

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test dashboard backend

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
> jenkins test dashboard backend

Test failure:

```2020-01-10 12:12:12,856.856 INFO:__main__:test_pwd_expiration_date_update (tasks.mgr.dashboard.test_user.UserTest) ... ERROR
2020-01-10 12:12:12,856.856 INFO:__main__:Running ['./bin/ceph', 'log', 'Ended test tasks.mgr.dashboard.test_user.UserTest.test_pwd_expiration_date_update']
2020-01-10 12:12:13,441.441 INFO:__main__:Stopped test: test_pwd_expiration_date_update (tasks.mgr.dashboard.test_user.UserTest) in 11.87283s
2020-01-10 12:12:13,442.442 INFO:__main__:
2020-01-10 12:12:13,443.443 INFO:__main__:======================================================================
2020-01-10 12:12:13,443.443 INFO:__main__:ERROR: test_pwd_expiration_date_update (tasks.mgr.dashboard.test_user.UserTest)
2020-01-10 12:12:13,443.443 INFO:__main__:----------------------------------------------------------------------
2020-01-10 12:12:13,443.443 INFO:__main__:Traceback (most recent call last):
2020-01-10 12:12:13,444.444 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/test_user.py", line 379, in test_pwd_expiration_date_update
2020-01-10 12:12:13,444.444 INFO:__main__:    self.login('user1', 'mypassword10#')
2020-01-10 12:12:13,444.444 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 87, in login
2020-01-10 12:12:13,444.444 INFO:__main__:    cls.logout()
2020-01-10 12:12:13,444.444 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 95, in logout
2020-01-10 12:12:13,445.445 INFO:__main__:    cls._post('/api/auth/logout')
2020-01-10 12:12:13,445.445 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 234, in _post
2020-01-10 12:12:13,445.445 INFO:__main__:    cls._request(url, 'POST', data, params)
2020-01-10 12:12:13,445.445 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 189, in _request
2020-01-10 12:12:13,445.445 INFO:__main__:    verify=False, headers=headers)
2020-01-10 12:12:13,446.446 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/sessions.py", line 572, in post
2020-01-10 12:12:13,446.446 INFO:__main__:    return self.request('POST', url, data=data, json=json, **kwargs)
2020-01-10 12:12:13,446.446 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/sessions.py", line 524, in request
2020-01-10 12:12:13,446.446 INFO:__main__:    resp = self.send(prep, **send_kwargs)
2020-01-10 12:12:13,446.446 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/sessions.py", line 637, in send
2020-01-10 12:12:13,446.446 INFO:__main__:    r = adapter.send(request, **kwargs)
2020-01-10 12:12:13,447.447 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/adapters.py", line 498, in send
2020-01-10 12:12:13,447.447 INFO:__main__:    raise ConnectionError(err, request=request)
2020-01-10 12:12:13,447.447 INFO:__main__:ConnectionError: ('Connection aborted.', error("(104, 'ECONNRESET')",))
2020-01-10 12:12:13,447.447 INFO:__main__:
2020-01-10 12:12:13,447.447 INFO:__main__:----------------------------------------------------------------------
2020-01-10 12:12:13,448.448 INFO:__main__:Ran 234 tests in 2329.236s
2020-01-10 12:12:13,448.448 INFO:__main__:
2020-01-10 12:12:13,448.448 INFO:__main__:FAILED (errors=1)
2020-01-10 12:12:13,448.448 INFO:__main__:
2020-01-10 12:12:13,449.449 INFO:__main__:======================================================================
2020-01-10 12:12:13,449.449 INFO:__main__:ERROR: test_pwd_expiration_date_update (tasks.mgr.dashboard.test_user.UserTest)
2020-01-10 12:12:13,449.449 INFO:__main__:----------------------------------------------------------------------
2020-01-10 12:12:13,449.449 INFO:__main__:Traceback (most recent call last):
2020-01-10 12:12:13,449.449 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/test_user.py", line 379, in test_pwd_expiration_date_update
2020-01-10 12:12:13,450.450 INFO:__main__:    self.login('user1', 'mypassword10#')
2020-01-10 12:12:13,450.450 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 87, in login
2020-01-10 12:12:13,450.450 INFO:__main__:    cls.logout()
2020-01-10 12:12:13,450.450 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 95, in logout
2020-01-10 12:12:13,450.450 INFO:__main__:    cls._post('/api/auth/logout')
2020-01-10 12:12:13,451.451 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 234, in _post
2020-01-10 12:12:13,451.451 INFO:__main__:    cls._request(url, 'POST', data, params)
2020-01-10 12:12:13,451.451 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 189, in _request
2020-01-10 12:12:13,451.451 INFO:__main__:    verify=False, headers=headers)
2020-01-10 12:12:13,451.451 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/sessions.py", line 572, in post
2020-01-10 12:12:13,451.451 INFO:__main__:    return self.request('POST', url, data=data, json=json, **kwargs)
2020-01-10 12:12:13,452.452 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/sessions.py", line 524, in request
2020-01-10 12:12:13,452.452 INFO:__main__:    resp = self.send(prep, **send_kwargs)
2020-01-10 12:12:13,452.452 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/sessions.py", line 637, in send
2020-01-10 12:12:13,452.452 INFO:__main__:    r = adapter.send(request, **kwargs)
2020-01-10 12:12:13,452.452 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/adapters.py", line 498, in send
2020-01-10 12:12:13,452.452 INFO:__main__:    raise ConnectionError(err, request=request)
2020-01-10 12:12:13,453.453 INFO:__main__:ConnectionError: ('Connection aborted.', error("(104, 'ECONNRESET')",))
```

**Comment by votdev** [comment]
> > jenkins test dashboard backend
> 
> Test failure:
> 
> ```
> 2020-01-10 12:12:12,856.856 INFO:__main__:Running ['./bin/ceph', 'log', 'Ended test tasks.mgr.dashboard.test_user.UserTest.test_pwd_expiration_date_update']
> 2020-01-10 12:12:13,441.441 INFO:__main__:Stopped test: test_pwd_expiration_date_update (tasks.mgr.dashboard.test_user.UserTest) in 11.87283s
> 2020-01-10 12:12:13,442.442 INFO:__main__:
> 2020-01-10 12:12:13,443.443 INFO:__main__:======================================================================
> 2020-01-10 12:12:13,443.443 INFO:__main__:ERROR: test_pwd_expiration_date_update (tasks.mgr.dashboard.test_user.UserTest)
> 2020-01-10 12:12:13,443.443 INFO:__main__:----------------------------------------------------------------------
> 2020-01-10 12:12:13,443.443 INFO:__main__:Traceback (most recent call last):
> 2020-01-10 12:12:13,444.444 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/test_user.py", line 379, in test_pwd_expiration_date_update
> 2020-01-10 12:12:13,444.444 INFO:__main__:    self.login('user1', 'mypassword10#')
> 2020-01-10 12:12:13,444.444 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 87, in login
> 2020-01-10 12:12:13,444.444 INFO:__main__:    cls.logout()
> 2020-01-10 12:12:13,444.444 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 95, in logout
> 2020-01-10 12:12:13,445.445 INFO:__main__:    cls._post('/api/auth/logout')
> 2020-01-10 12:12:13,445.445 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 234, in _post
> 2020-01-10 12:12:13,445.445 INFO:__main__:    cls._request(url, 'POST', data, params)
> 2020-01-10 12:12:13,445.445 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 189, in _request
> 2020-01-10 12:12:13,445.445 INFO:__main__:    verify=False, headers=headers)
> 2020-01-10 12:12:13,446.446 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/sessions.py", line 572, in post
> 2020-01-10 12:12:13,446.446 INFO:__main__:    return self.request('POST', url, data=data, json=json, **kwargs)
> 2020-01-10 12:12:13,446.446 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/sessions.py", line 524, in request
> 2020-01-10 12:12:13,446.446 INFO:__main__:    resp = self.send(prep, **send_kwargs)
> 2020-01-10 12:12:13,446.446 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/sessions.py", line 637, in send
> 2020-01-10 12:12:13,446.446 INFO:__main__:    r = adapter.send(request, **kwargs)
> 2020-01-10 12:12:13,447.447 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/adapters.py", line 498, in send
> 2020-01-10 12:12:13,447.447 INFO:__main__:    raise ConnectionError(err, request=request)
> 2020-01-10 12:12:13,447.447 INFO:__main__:ConnectionError: ('Connection aborted.', error("(104, 'ECONNRESET')",))
> 2020-01-10 12:12:13,447.447 INFO:__main__:
> 2020-01-10 12:12:13,447.447 INFO:__main__:----------------------------------------------------------------------
> 2020-01-10 12:12:13,448.448 INFO:__main__:Ran 234 tests in 2329.236s
> 2020-01-10 12:12:13,448.448 INFO:__main__:
> 2020-01-10 12:12:13,448.448 INFO:__main__:FAILED (errors=1)
> 2020-01-10 12:12:13,448.448 INFO:__main__:
> 2020-01-10 12:12:13,449.449 INFO:__main__:======================================================================
> 2020-01-10 12:12:13,449.449 INFO:__main__:ERROR: test_pwd_expiration_date_update (tasks.mgr.dashboard.test_user.UserTest)
> 2020-01-10 12:12:13,449.449 INFO:__main__:----------------------------------------------------------------------
> 2020-01-10 12:12:13,449.449 INFO:__main__:Traceback (most recent call last):
> 2020-01-10 12:12:13,449.449 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/test_user.py", line 379, in test_pwd_expiration_date_update
> 2020-01-10 12:12:13,450.450 INFO:__main__:    self.login('user1', 'mypassword10#')
> 2020-01-10 12:12:13,450.450 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 87, in login
> 2020-01-10 12:12:13,450.450 INFO:__main__:    cls.logout()
> 2020-01-10 12:12:13,450.450 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 95, in logout
> 2020-01-10 12:12:13,450.450 INFO:__main__:    cls._post('/api/auth/logout')
> 2020-01-10 12:12:13,451.451 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 234, in _post
> 2020-01-10 12:12:13,451.451 INFO:__main__:    cls._request(url, 'POST', data, params)
> 2020-01-10 12:12:13,451.451 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 189, in _request
> 2020-01-10 12:12:13,451.451 INFO:__main__:    verify=False, headers=headers)
> 2020-01-10 12:12:13,451.451 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/sessions.py", line 572, in post
> 2020-01-10 12:12:13,451.451 INFO:__main__:    return self.request('POST', url, data=data, json=json, **kwargs)
> 2020-01-10 12:12:13,452.452 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/sessions.py", line 524, in request
> 2020-01-10 12:12:13,452.452 INFO:__main__:    resp = self.send(prep, **send_kwargs)
> 2020-01-10 12:12:13,452.452 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/sessions.py", line 637, in send
> 2020-01-10 12:12:13,452.452 INFO:__main__:    r = adapter.send(request, **kwargs)
> 2020-01-10 12:12:13,452.452 INFO:__main__:  File "/tmp/tmp.YPHoMV6fOL/venv/local/lib/python2.7/site-packages/requests/adapters.py", line 498, in send
> 2020-01-10 12:12:13,452.452 INFO:__main__:    raise ConnectionError(err, request=request)
> 2020-01-10 12:12:13,453.453 INFO:__main__:ConnectionError: ('Connection aborted.', error("(104, 'ECONNRESET')",))
> ```

Can not reproduce it on my local machine and the error does not look related to this PR, so i'll trigger another run.

**Comment by votdev** [comment]
jenkins test dashboard backend

**Comment by callithea** [comment]
> Can not reproduce it on my local machine and the error does not look related to this PR, so i'll trigger another run.

Thanks for the update!

**Comment by votdev** [comment]
jenkins test dashboard backend
